### PR TITLE
{ARM} Bump `azure-mgmt-resource` to `21.1.0b1` and refine processing logic for JSONC template

### PR DESCRIPTION
**Pull request by zhoxing-ms**
**Description**

Feature request - Close: #22073
Refer to #22030 and replace it

The SDK package `azure-mgmt-resource` has a breaking change after version `21.1.0`:
1. In the original SDK version, JSON serialization exception will be suppressed silently [code link](https://github.com/Azure/azure-sdk-for-python/blob/77a1abaea4d2bb07e3abf20020e82dc7195a5fe3/sdk/core/azure-core/azure/core/pipeline/transport/_base.py#L557-L560)
```Python
try:
    request.set_json_body(content)
except TypeError:
    request.data = content
```

2. In the new SDK version, JSON serialization errors will be thrown directly [code link](https://github.com/Azure/azure-sdk-for-python/blob/73f4c8fb284a593a8fec4381e423b05c543172cf/sdk/core/azure-core/azure/core/rest/_rest_py3.py#L142-L144)
```Python
if json is not None:
    default_headers, self._data = set_json_body(json)
    return default_headers
```

Because ARM template does not support normal JSON, but JSONC, the JSON serialization exception `Object of type JsonCTemplate is not JSON serializable` will appear in the new SDK version.

Therefore, we refined `JsonCTemplate` to add compatible logic to the new SDK package:
After removing `JSONSerializer` and `JsonCTemplate`, `json['properties']['template']` will not be the format of non JSON object when it is serialized in `_set_body()`, and the subsequent serialization exception will not appear

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ARM

**Comment by zhoxing-ms** [comment]
https://github.com/Azure/azure-sdk-for-python/blob/73f4c8fb284a593a8fec4381e423b05c543172cf/sdk/core/azure-core/azure/core/rest/_rest_py3.py#L142-L144

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 1]
```diff
@@ -388,41 +382,27 @@ def _deploy_arm_template_core_unmodified(cmd, resource_group_name, template_file
...
-    def body(self, data, data_type, **kwargs):
-        if data_type in ('Deployment', 'ScopedDeployment', 'DeploymentWhatIf', 'ScopedDeploymentWhatIf'):
-            # Be sure to pass a DeploymentProperties
-            template = data.properties.template
-            if template:
-                data_as_dict = data.serialize()
-                data_as_dict["properties"]["template"] = JsonCTemplate(template)
-
-                return data_as_dict
-        return super(JSONSerializer, self).body(data, data_type, **kwargs)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:406]: After removing the customized serialization, the information of the ARM template can also be printed in the debug log (only the parameter names and values were printed before)

Before
```
{'properties': {'template': <azure.cli.command_modules.resource.custom.JsonCTemplate object at 0x000002020F2EE970>, 'parameters': {'location': {'value': 'westus'}, 'name': {'value': 'azure-cli-deploy-test-nsg1'}}, 'mode': 'Incremental'}}
```

After
```
{"properties": {"template": "{\r\n    \"$schema\": \"https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#\",\r\n    \"contentVersion\": \"1.0.0.0\",\r\n    \"parameters\": {\r\n      \"location\": {\r\n        \"type\": \"string\",\r\n        \"defaultValue\": \"[resourceGroup().location]\",\r\n        \"metadata\": {\r\n          \"description\": \"Location for the network security group.\\r\\n\"\r\n        }\r\n      },\r\n      \"name\": {\r\n        \"type\": \"string\",\r\n        \"metadata\": {\r\n          \"description\": \"Name of the network security group.\\r\\n\"\r\n        }\r\n      }\r\n    },\r\n    \"variables\": {\r\n    },\r\n    \"resources\": [\r\n      {\r\n        \"type\": \"Microsoft.Network/networkSecurityGroups\",\r\n        \"name\": \"[parameters('name')]\",\r\n        \"apiVersion\": \"2015-06-15\",\r\n        \"location\": \"[parameters('location')]\",\r\n        \"properties\": {\r\n          \"securityRules\": [\r\n          ]\r\n        },\r\n        \"dependsOn\": [ ]\r\n      }\r\n    ],\r\n    \"outputs\": {\r\n      \"NewNSG\": {\r\n        \"type\": \"object\",\r\n        \"value\": \"[reference(parameters('name'))]\"\r\n      }\r\n    }\r\n}", "parameters": {"location": {"value": "westus"}, "name": {"value": "azure-cli-deploy-test-nsg1"}}, "mode": "Incremental"}}
```

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:406]: @alex-frankel Do you think it is acceptable to print the ARM template in the CLI debug log?

**Review by jsntcy — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
@cxznmhdcxz  has helped complete the tests related to ARM deployment in live mode locally

**Pull request merged**